Rationals added to fancysets

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
closes #9609 with added tests
closes #10757 by adding Rationals
closes #11863
closes #16268 by defining interaction for Naturals and Reals
closes #16938 
closes #16940 

#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- sets
    - Rationals added to fancysets
    - interactions between Z, N, N0 and R is improved
<!-- END RELEASE NOTES -->

oscarbenjamin commented:
Rationals here doesn't print nicely:
```
In [1]: S.Rationals                                                                                                               
Out[1]: Rationals

In [2]: S.Reals                                                                                                                   
Out[2]: ℝ

In [3]: S.Integers                                                                                                                
Out[3]: ℤ
```
What is `Rationals.__iter__` useful for? It currently only gives rational numbers between +-1. There is a way to enumerate all rationals in the diagram here though:
https://en.wikipedia.org/wiki/Rational_number#Properties

oscarbenjamin commented on sympy/sets/fancysets.py:
```diff
@@ -9,9 +9,56 @@
+        d = 2
+        while True:
+            for n in range(d):
+                if igcd(n, d) == 1:
```
This misses the integers

smichr replied:
The integers are printed as `d/n` so `n,d`=`(2,4)` is skipped but (1,2) isn't and the (1,2) emits `1/2,2,-1/2,-2`. i.e., all integers are emitted when n is 1.

smichr commented:
This result is driving me crazy...I will have to step through and see why 
`S.Naturals.intersect(S.Reals) -> Range(1, oo, 1)` but `S.Naturals0` gives itself.

smichr commented:
```console
def e2d(s):
    if type(s) is str:
        s = s.replace('==', '=')
        return [tuple(map(lambda i:i.strip(), li.split('='))) for li in s.strip().splitlines()]
    elif isinstance(s, Eq):
        return s.args
    elif type(s) is list:
        return list(map(e2d, s))

e2d('v_f=x\nsigma_f=f\nsigma_c=c\nsigma_p=p')

s2d = lambda s: [tuple(map(lambda i:i.strip(), li.split('='))) for li in s.strip().splitlines()]
```

oscarbenjamin commented on sympy/printing/pretty/pretty_symbology.py:
```diff
@@ -501,6 +501,7 @@ def xsym(sym):
                                 (U('DOUBLE-STRUCK CAPITAL N') +
                                  U('SUBSCRIPT ZERO'))),
     'Integers':                U('DOUBLE-STRUCK CAPITAL Z'),
+    'Rationals':               U('DOUBLE-STRUCK CAPITAL Q'),
```
I don't see the pretty result here:
```
In [1]: S.Reals                                                                                                                   
Out[1]: ℝ

In [2]: S.Rationals                                                                                                               
Out[2]: Rationals
```

smichr replied:
I'm a neophyte at pretty printing. Any ideas?

oscarbenjamin replied:
Something like this (although with a better ASCII fallback):
```diff
diff --git a/sympy/printing/pretty/pretty.py b/sympy/printing/pretty/pretty.py
index 6b9dd8c3f2..9c5802117c 100644
--- a/sympy/printing/pretty/pretty.py
+++ b/sympy/printing/pretty/pretty.py
@@ -172,6 +172,13 @@ def _print_Reals(self, e):
             inf_list = ['-oo', 'oo']
             return self._print_seq(inf_list, '(', ')')
 
+    def _print_Rationals(self, e):
+        if self._use_unicode:
+            return self._print_Atom(e)
+        else:
+            inf_list = ['-oo', 'oo']
+            return self._print_seq(inf_list, '(', ')')
+
     def _print_subfactorial(self, e):
         x = e.args[0]
         pform = self._print(x)
```
That gives:
```julia
In [1]: S.Rationals                                                                                                               
Out[1]: ℚ
```

oscarbenjamin commented on sympy/sets/fancysets.py:
```diff
@@ -9,9 +9,54 @@
+    True
+    >>> iterable = iter(S.Rationals)
+    >>> [next(iterable) for i in range(10)]
+    [0, 1/2, 2, -1/2, -2, 1/3, 3, -1/3, -3, 2/3]
```
Will this ever yield 1 and -1?

oscarbenjamin commented on sympy/sets/fancysets.py:
```diff
@@ -9,9 +9,54 @@
+
+    def __iter__(self):
+        from sympy.core.numbers import igcd, Rational
+        yield S.Zero
```
Maybe yield 1 and -1 here

oscarbenjamin commented on sympy/sets/tests/test_fancysets.py:
```diff
@@ -518,6 +603,16 @@ def test_ImageSet_simplification():
+    raises(ValueError, lambda:
+        {1, 2} in imageset(x, (x, 1/x), S.Integers))
+    assert (1, 2) not in imageset(x, x + 1, S.Integers)
+    assert {1} in imageset(x, {x}, S.Integers)
```
Why should `{1, 2}` raise?

smichr replied:
`{1, 2}` is a set, not an Integer; `S.Integers.intersect({1,2}) == {1, 2}`, however.

oscarbenjamin replied:
Should it not be False then since `{1, 2}` is not an element of the set? In general a set can be an element of another set so I don't see why it needs to raise

smichr replied:
Agree. My thinking was that we should be more careful with ImageSet otherwise we let something fail silently that didn't make sense for the query (like `Eq(I, 0) -> False vs Lt(I, 0) -> error`). In my answer, even, I got it wrong. {1, 2} fails because a set will not be in a set of tuples, not because it is not an Integer. `1 in imageset(x, (x, 1/x), S.Integers)` also fails.

Will change.

smichr committed:
16938: check Lambda for arg dups

smichr committed:
add Rationals to fancysets

smichr committed:
fix integer itersections with Interval

smichr committed:
16940: unevaluated Intersection from imageset

smichr committed:
function interactions with Z,N,N0

smichr committed:
16878: imageset with Real base

smichr committed:
modify coordsysrect test

smichr commented:
@oscarbenjamin , do you want to give this a final check?

smichr committed:
closes #11863

smichr committed:
q > 0 so abs(q) == q

oscarbenjamin commented on sympy/sets/tests/test_fancysets.py:
```diff
@@ -802,3 +811,31 @@ def test_issue_16871b():
+    assert 2 in S.Rationals
+    r = symbols('r', rational=True)
+    assert r in S.Rationals
+    raises(ValueError, lambda: x in S.Rationals)
```
I get SympifyError here:
```
>>> from sympy import S, Symbol
>>> x = Symbol('x')
>>> x in S.Rationals
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/enojb/current/sympy/sympy/sympy/sets/sets.py", line 561, in __contains__
    symb = sympify(self.contains(other))
  File "/Users/enojb/current/sympy/sympy/sympy/sets/sets.py", line 288, in contains
    ret = sympify(self._contains(other))
  File "/Users/enojb/current/sympy/sympy/sympy/sets/fancysets.py", line 45, in _contains
    return _sympify(other.is_rational)
  File "/Users/enojb/current/sympy/sympy/sympy/core/sympify.py", line 417, in _sympify
    return sympify(a, strict=True)
  File "/Users/enojb/current/sympy/sympy/sympy/core/sympify.py", line 279, in sympify
    raise SympifyError(a)
sympy.core.sympify.SympifyError: SympifyError: None
```
Is that an intentional exception?

That also happens with `S.Rationals.contains(x)`. Does contains have to return True/False or is None a possibility there?

smichr replied:
The whole sets module needs a good audit for what must/can be returned from each method. Like should x in set raise an error if it is indeterminate? What about set.contains(x)? It seems like the latter should allow Contains as a result. I am not sure of the former.

oscarbenjamin replied:
We can rule out some possibilities since `x in y` has to be a bool. Demo:
```python
class NotSure:
    def __bool__(self):
        return self

class Equivocal:
    def __contains__(self, other):
        return NotSure()

x in Equivocal()
```
Result:
```
TypeError: __bool__ should return bool, returned NotSure
```
So `x in y` only has 3 possibilities: True, False or exception.

oscarbenjamin commented on sympy/sets/tests/test_sets.py:
```diff
@@ -14,17 +14,30 @@
+    assert imageset(x, abs(x), S.Integers) is S.Naturals0
+    # issue 16878a
+    r = symbols('r', real=True)
+    assert (1, r) in imageset(x, (x, x), S.Reals)
```
Maybe I misunderstand: this is the set `(x,x)` so `(1,r)` is only in the set if `r=1` and we don't know that r is 1...

smichr replied:
hmm...agree

oscarbenjamin commented on sympy/sets/tests/test_sets.py:
```diff
@@ -14,17 +14,30 @@
-    assert imageset(x, y, ints) == FiniteSet(y)
+    assert imageset(x, y, ints) == Intersection({y}, ints)
+    assert imageset((x, y), (1, z), ints*S.Reals) == Intersection(
+        {(1, z)}, ints*S.Reals)
```
Should the result here not just be `{(1,z)}`?

smichr replied:
not unless z is real

oscarbenjamin replied:
`ints*S.Reals` is the domain not the codomain e.g
```
In [11]: imageset(x, S.Half, S.Integers)                                                                                          
Out[11]: ∅
```
er... wait that is not what I expected. How should I interpret this?

smichr replied:
ugh...this should just be {1/2}: plug in an int always get out 1/2. So then `imageset(x, z, Integers)` must also be `{z}`.

smichr committed:
remove use of .is_EmptySet; can now be deprecated

smichr committed:
clean up contains and as_relational

smichr committed:
16940; imageset expr independent of variables

smichr committed:
clean up imageset._contains

smichr committed:
return False instead of raising ValueErrors

smichr commented:
I am checking coverage.

smichr committed:
use try/except for as-relational

smichr committed:
restore evaluate=True to allow evalulation

smichr committed:
allow for TypeError, too

smichr commented:
@oscarbenjamin , besides fixing some of the broken sets issues, the added functionality is `Rationals` and the changed functionality is that `contains` returns the `as_relational` form when it makes sense to do so. This ordered variables and symbols differently than the construction of the same, so tests were changed for that reason.

oscarbenjamin commented on sympy/sets/tests/test_sets.py:
```diff
@@ -14,17 +14,31 @@
+    assert imageset(x, abs(x), S.Integers) is S.Naturals0
+    # issue 16878a
+    r = symbols('r', real=True)
+    assert (1, r) not in imageset(x, (x, x), S.Reals)
```
I think this one should raise

oscarbenjamin commented:
This looks good. I left one comment but that can be punted to another issue if this is merged.
